piptools/utils.py

```diff
@@ -695,11 +686,27 @@ def parse_config_file(config_file: Path) -> dict[str, Any]:
             hint=f"Could not parse '{config_file !s}': {value_err !s}",
         )
 
+    cli_opts = {
+        opt: option
+        for option in click_context.command.params
+        for opt in itertools.chain(option.opts, option.secondary_opts)
+    }
+
     # In a TOML file, we expect the config to be under `[tool.pip-tools]`
     piptools_config: dict[str, Any] = config.get("tool", {}).get("pip-tools", {})
-    piptools_config = {
-        get_click_dest_for_option(k): v for k, v in piptools_config.items()
-    }
+
+    # Replace boolean flags like ``--no-annotate`` with their equivalents
+    try:
+        piptools_config = {
+            cli_opts["--" + k].name: (
+                not v
+                if k.startswith("no-") and not k == "no-index" and isinstance(v, bool)
```

Comment:
We can omit `not k == "no-index"` condition, because `cli_opts["--no-index"]` will raise KeyError

Reply:
Sadly it doesn't, and breaks the tests. I've added this condition after the failures. So far I've not found a way to properly detect a `click` boolean flag.

Reply:
And with this code https://github.com/jazzband/pip-tools/pull/1913/files#r1264749661 we don't need  `not k == "no-index"` check too.